src/GraphQL.Tests/Types/ComplexGraphTypeTests.cs

```diff
@@ -347,8 +347,10 @@ public void throws_when_field_name_doesnot_follow_spec(string fieldName)
         [InlineData("id$")]
         public void does_not_throw_with_filtering_nameconverter(string fieldName)
         {
+            NameValidator.Validation = (n, t) => { };
             var type = new ObjectGraphType();
             type.Field<StringGraphType>(fieldName);
+            NameValidator.Validation = NameValidator.ValidateDefault;
```

Comment:
I do not think this solution is a satisfactory answer.  Would the user really set it back after they initialized the graph?  If they did it as shown in the sample here with a scoped schema, multiple threads would be modifying the `NameValidator.Validation` property simultaneously, for sure breaking their code.  If they didn't, the schema would not be properly verified.  And it should be, even when using a custom `NameValidator`.

Reply:
:)  I guessed that these lines might raise questions. In fact, these calls are only made to save your test, no more. No need from client to set back static property. I called ` NameValidator.Validation = (n, t) => { };` to allow setting _bad_ name in line `  type.Field<StringGraphType>(fieldName);` . Then I set it back to show that your `TestNameConverter` works. It changes field name from _bad_ one to _good_ one. That is all.

Reply:
Well that's why I didn't put it in my PR.

I think:
- The NameConverter needs the ability to convert bad names to good ones
- The schema should be validated after that happens.

These changes invalidate that.

Reply:
I suggest a boolean that disables pre-initialization checks, retaining the checks during initialization.

Reply:
> The NameConverter needs the ability to convert bad names to good ones

So it happens.

> The schema should be validated after that happens.

This validation is not enough now until we freeze the whole schema. We now allow to change almost any part of the schema at any time. If we change the design in future, then of course it will be possible to get rid of these checks inside setters.

> I suggest a boolean that disables pre-initialization checks, retaining the checks during initialization.

What does it look like? `NameValidator.Validation = (n, t) => { };` is some sort of boolean already.

Reply:
In general, now everything works as before, with the difference that there are fewer options to make a error in runtime.

Reply:
The difference is that the boolean doesn't disable the check during schema initialization.

Reply:
An equivalent suggestion would be to leave all the code you wrote as-is, except to change the calls within `HandleField` to call `ValidateDefault` rather than `Validate`.

Reply:
That is, do you propose to tighten checks?

Reply:
Compared to this current PR, yes.  Compared to current codebase, no.

Reply:
> except to change the calls within HandleField to call ValidateDefault rather than Validate

I don't understand how to do this yet. `NameValidator` is static 'low-level' class. I saw it as a global switch that users turn on at their own discretion. The library itself doesn't manipulate it.

> I suggest a boolean that disables pre-initialization checks

What can we achieve by disabling pre-initialization checks?

Reply:
> What can we achieve by disabling pre-initialization checks?

Retaining proper validation of the initialized schema, while allowing a NameConverter to fix up broken names.  In my mind, this is half of the goal of #1840 .  Allowing a user to bypass ALL validation doesn't fit with that.

Separately if we have some way of bypassing ALL validation (e.g. for #1451) … ok … but we shouldn't get rid of our normal validation checks just because we want to add that feature.

Reply:
To be honest I think that NameConverter never used for _fix up broken names_. It is used to follow conventions - PascalCase, camelCase, snake_case and so on.

> Allowing a user to bypass ALL validation doesn't fit with that.

I share this concern. But we have already found out that we want to give the opportunity to do so in exceptional cases. I can specifically write a separate paragraph in the generated documentation.

Regarding #1451 . Two new introspection types bypass validation by means of internal methods. The same methods have long been used for other types of instrospection. I'm talking about `internal void GraphType.SetName(string name, bool validate)`.

> but we shouldn't get rid of our normal validation checks 

It seems that we do not get rid of them actually. What is _normal validation checks_ in your opinion? I honestly don't see anything _abnormal_ in the checks if this PR. Everything (checks, name converters) works by default. If (and only if) someone wants to bypass scrict checks for names, well ... OK, he has the possibility to do so _after being warned in advance_.

Although I have a suspicion that I see only part of the picture, therefore I do not understand your concerns.

Reply:
I'm not sure what there is to explain.  Most likely, I see my viewpoint as clearly as you see yours - lol.  It seems like we are each focusing on completely different issues here, and likely have some different viewpoints on those issues.

1. You are trying to allow a way to bypass validation for invalid names

I don't have a problem with this.  But I think that bypassing validation has nothing to do with moving the checks from `GraphTypesLookup.HandleField` to `FieldType.Name` and `QueryArgument.Name`.  If you want to write a PR for bypassing validation, then do so.  But if the validation is not bypassed, then the name should be checked after the `INameConverter` is applied, and not before.

2. You are trying to consolidate checks

Conceptually I agree, but they are already consolidated.  Names are only validated within `HandleField.GraphTypesLookup`.  There are basic null checks where it makes sense, of course.  I don't see how you can consolidate something that's already consolidated.

Futher, if we need more than one check to fulfill the criteria here, so be it.  I wouldn't sacrifice code quality for a couple line of code.

3. We would both prefer if the checks were within the property setters

I agree.... if it can be done within the confines of the other criteria here.

4. I'm trying to enforce a valid schema (unless someone specifies otherwise, if we provide that option), regardless of whether they use a `INameConverter` or not.

I think we have real, valid cases of people wanting to use a name converter to fix up names that may be invalid.  See #1840 and https://github.com/graphql-dotnet/graphql-dotnet/issues/1949#issuecomment-717328587 .  And, it just makes sense.  No validation on names after being converted - like really?  The name converter is a main feature of the program, used by almost every schema out there.  True, the default one won't generate invalid names (if not fed invalid names), but this isn't some interface that was haphazardly added to allow some crazy scenario.

It would seem that you don't feel that this is a valuable feature / bugfix.  At least, you would sacrifice this for points 2 and 3 above.

Reply:
> Regarding #1451 . Two new introspection types bypass validation by means of internal methods. The same methods have long been used for other types of instrospection. I'm talking about `internal void GraphType.SetName(string name, bool validate)`.

Unless it can be implemented in a way that does not break spec, I think we should explore a way to introduce this as a separate library.  This would also be a valuable sample of how third-parties could add similar out-of-spec features to the core library.  So perhaps we should simply expose `SetName` as a public member, so anyone can set a name and bypass validation.  (Would require changes to `HandleField` as it is now, but not with this PR.)

Reply:
Regarding https://github.com/graphql-dotnet/graphql-dotnet/issues/1949#issuecomment-717328587 . `NameConverter` does not help since it works with arg names and field names, not type names and enum value names.

Reply:
I'm OK with periodically merging master into my #1451 fork. It is difficult to move it into a separate library.